cluster-essentials/cluster_init.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package clusteressential has APIs for bootstrapping kubernetes cluster
+package clusteressential
+
+import (
+	"bytes"
+	"os"
+
+	"github.com/cppforlife/go-cli-ui/ui"
+	"github.com/pkg/errors"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/vmware-tanzu/carvel-imgpkg/pkg/imgpkg/cmd"
+	"github.com/vmware-tanzu/tanzu-framework/util/kube"
+)
+
+// FUNCTION IS NOT COMPLETE
+func generateTemplateFiles(component string) (manifestFileName string, err error) {
+	// TODO generate template file using ytt, in case of overlay
+	fileName := "/tmp/" + component + ".yaml"
+	return fileName, nil
+}
+
+func downloadClusterEssentialBundle(clusterEssentialBundlePath string) error {
+	var outputBuf, errorBuf bytes.Buffer
+	writerUI := ui.NewWriterUI(&outputBuf, &errorBuf, nil)
+	outputDir := "tmp"
+	pullOptions := cmd.NewPullOptions(writerUI)
+	pullOptions.OutputPath = outputDir
+	pullOptions.ImageFlags = cmd.ImageFlags{Image: clusterEssentialBundlePath}
+	err := pullOptions.Run()
+	return err
+}
+
+func InitClusterEssentials(clusterKubeconfigPath, clusterEssentialBundlePath string) error {
+	config, err := clientcmd.LoadFromFile(clusterKubeconfigPath)
+	if err != nil {
+		return errors.Wrapf(err, "failed to load kubeconfig from %s", clusterKubeconfigPath)
+	}
+	rawConfig, err := clientcmd.Write(*config)
+	if err != nil {
+		return errors.Wrap(err, "unable to write config")
+	}
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(rawConfig)
+	if err != nil {
+		return errors.Wrap(err, "unable to set up rest config")
+	}
+	dynamicClient, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return errors.Wrap(err, "unable to set up rest config")
+	}
+	err = downloadClusterEssentialBundle(clusterEssentialBundlePath)
+	if err != nil {
+		return errors.Wrap(err, "unable to download cluster essential manifest")
+	}
+	components := [2]string{"kapp-controller", "secretgen-controller"}
+	for _, component := range components {
+		fileName, err := generateTemplateFiles(component)
+		if err != nil {
+			return errors.Wrapf(err, "unable to generate template")
+		}
+		input, err := os.ReadFile(fileName)
+		if err != nil {
+			return errors.Wrapf(err, "unable to read file %s", fileName)
+		}
+		generatedManifestBytes := []byte(string(input))
+		err = kube.CreateResourcesFromManifest(generatedManifestBytes, restConfig, dynamicClient)
```

Comment:
This function will invoke an equivalent of the `kubectl create` command which is not an idempotent operation. Lets say the `secretgen-controller` install fails due to some reason, and user triggers the cluster init flow again after fixing it, the kapp-controller `create` operation will fail, since resource already exists. Why not use `kubectl apply` equivalent instead?